### PR TITLE
Refactor cult dagger rune drawing and null rod rune clearing into components + minor rune / blood magic tweaks

### PR DESCRIPTION
**Pull request by MrMelbert**
## About The Pull Request

- The act of inscribing runes, anchoring / unanchoring cult structures, deconstructing cult girders, and purging cultists of holy water have all been moved off of the cult dagger itself and onto a component, `cult_ritual_item`. 
- The act of destroying a cult rune or heretic rune with the null rod have been moved off of their respective items and onto the nullrod itself.
  - Added a signal, `COMSIG_ITEM_ATTACK_EFFECT`, for the above purposes. 
- Various cult spells and dagger interactions have had their code improved slightly. 

- [ ] Test no regression / accidental gameplay changes
- [x] Test the null rod works as intended. 

## Why It's Good For The Game

Cult dagger stuff was all over the place, same with null rod cleansing. This brings it all together and unified the behavior, also allowing more than daggers to inscribe runes and work with cult items eventually. 

## Changelog

:cl: Melbert
refactor: Refactored cult rune inscribing into a component
refactor: Refactored null rod rune clearing into a component
/:cl:

### CODE REVIEW

File: code/modules/antagonists/cult/ritual.dm [hunk 1]
```diff
@@ -1,164 +0,0 @@
...
-
-*/
-
-
-/obj/item/melee/cultblade/dagger/Initialize(mapload)
-	. = ..()
-	if(!LAZYLEN(GLOB.rune_types))
-		GLOB.rune_types = list()
-		var/static/list/non_revealed_runes = (subtypesof(/obj/effect/rune) - /obj/effect/rune/malformed)
-		for(var/i_can_do_loops_now_thanks_remie in non_revealed_runes)
```
Comment (MrMelbert) [inline code/modules/antagonists/cult/ritual.dm:13]: :pensive:

File: code/datums/components/cult_ritual_item.dm [hunk 2]
```diff
@@ -0,0 +1,465 @@
...
+		COMSIG_PARENT_EXAMINE,
+		))
+
+/*
+ * Signal proc for [COMSIG_PARENT_EXAMINE].
+ * Gives the examiner, if they're a cultist, our set examine message.
+ * Usually, this will include various instructions on how to use the thing.
+ */
+/datum/component/cult_ritual_item/proc/on_examine(datum/source, mob/examiner, list/examine_text)
+	if(!IS_CULTIST(examiner))
```
Comment (Watermelon914) [inline code/datums/components/cult_ritual_item.dm:77]:
```suggestion
	SIGNAL_HANDLER
	if(!IS_CULTIST(examiner))
```

File: code/datums/components/cult_ritual_item.dm [hunk 3]
```diff
@@ -0,0 +1,465 @@
...
+		return
+
+	examine_text += examine_message
+
+/*
+ * Signal proc for [COMSIG_ITEM_ATTACK_SELF].
+ * Allows the user to begin scribing runes.
+ */
+/datum/component/cult_ritual_item/proc/try_scribe_rune(datum/source, mob/user)
+	if(!isliving(user))
```
Comment (Watermelon914) [inline code/datums/components/cult_ritual_item.dm:87]:
```suggestion
	SIGNAL_HANDLER
	if(!isliving(user))
```

File: code/datums/components/cult_ritual_item.dm [hunk 4]
```diff
@@ -0,0 +1,465 @@
...
+
+	INVOKE_ASYNC(src, .proc/do_purge_holywater, user)
+
+
+/*
+ * Signal proc for [COMSIG_ITEM_ATTACK_OBJ].
+ * Allows the ritual items to unanchor cult buildings or destroy rune girders.
+ */
+/datum/component/cult_ritual_item/proc/try_hit_object(datum/source, obj/structure/target, mob/cultist)
+	if(!isliving(cultist) || !IS_CULTIST(cultist))
```
Comment (Watermelon914) [inline code/datums/components/cult_ritual_item.dm:123]:
```suggestion
	SIGNAL_HANDLER
	if(!isliving(cultist) || !IS_CULTIST(cultist))
```

File: code/datums/components/cult_ritual_item.dm [hunk 5]
```diff
@@ -0,0 +1,465 @@
...
+	if(istype(target, /obj/structure/destructible/cult))
+		INVOKE_ASYNC(src, .proc/do_unanchor_structure, target, cultist)
+		return COMPONENT_NO_AFTERATTACK
+
+/*
+ * Signal proc for [COMSIG_ITEM_ATTACK_EFFECT].
+ * Allows the ritual items to remove runes.
+ */
+/datum/component/cult_ritual_item/proc/try_clear_rune(datum/source, obj/effect/target, mob/living/cultist, params)
+	if(!isliving(cultist) || !IS_CULTIST(cultist))
```
Comment (Watermelon914) [inline code/datums/components/cult_ritual_item.dm:139]:
```suggestion
	SIGNAL_HANDLER
	if(!isliving(cultist) || !IS_CULTIST(cultist))
```

File: code/datums/components/effect_remover.dm [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+	RegisterSignal(parent, COMSIG_ITEM_ATTACK_EFFECT, .proc/try_remove_effect)
+
+/datum/component/effect_remover/UnregisterFromParent()
+	UnregisterSignal(parent, COMSIG_ITEM_ATTACK_EFFECT)
+
+/*
+ * Signal proc for [COMSIG_ITEM_ATTACK_EFFECT].
+ */
+/datum/component/effect_remover/proc/try_remove_effect(datum/source, obj/effect/target, mob/living/user, params)
+	if(!isliving(user))
```
Comment (Watermelon914) [inline code/datums/components/effect_remover.dm:45]:
```suggestion
	SIGNAL_HANDLER
	if(!isliving(user))
```

**Review by Watermelon914 — CHANGES_REQUESTED** [review]
Make sure to also check your signal procs because I might've missed one

File: code/datums/components/cult_ritual_item.dm [hunk 7]
```diff
@@ -0,0 +1,474 @@
...
+
+	if(!isitem(parent))
+		return COMPONENT_INCOMPATIBLE
+
+	// Initialize all our rune types if it wasn't already done.
+	if(!LAZYLEN(GLOB.rune_types))
+		for(var/obj/effect/rune/rune as anything in subtypesof(/obj/effect/rune))
+			if(!initial(rune.can_be_scribed))
+				continue
+			LAZYSET(GLOB.rune_types, initial(rune.cultist_name), rune) // Uses the cultist name for displaying purposes
```
Comment (Watermelon914) [inline code/datums/components/cult_ritual_item.dm:35]: this should really be done when the rune_types list is initialized

**Review by Watermelon914 — CHANGES_REQUESTED** [review]

File: code/datums/components/cult_ritual_item.dm (hunk 7 above)
Comment (MrMelbert) [reply code/datums/components/cult_ritual_item.dm:35]: In the past it didn't initialize rune types unless a ritual dagger was initialized, presumably to save memory not filling a list that wouldn't be used? I can't change that though.

Comment (Watermelon914) [reply code/datums/components/cult_ritual_item.dm:35]: why not? you can run procs when the list is initialized

Comment (MrMelbert) [reply code/datums/components/cult_ritual_item.dm:35]: My bad, I meant to say "I can change that". Will do.

File: code/datums/components/cult_ritual_item.dm [hunk 8]
```diff
@@ -0,0 +1,463 @@
...
+		return FALSE
+
+	return TRUE
+
+/*
+ * Removes all shields from the shields list.
+ */
+/datum/component/cult_ritual_item/proc/cleanup_shields()
+	for(var/obj/structure/emergency_shield/cult/narsie/shield as anything in shields)
+		LAZYREMOVE(shields, shield)
```
Comment (Watermelon914) [inline code/datums/components/cult_ritual_item.dm:461]:
```suggestion
		LAZYREMOVE(shields, shield)
```
instead, just do LAZYCLEARLIST(shields) at the end of the proc, it's more optimal
Additionally you can do QDEL_LAZYLIST(shields)

Comment (Watermelon914) [reply code/datums/components/cult_ritual_item.dm:461]: It's not a big deal at all though so I'll just be merging this anyways

**Pull request merged**